digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/service/AwsService.scala

```diff
@@ -0,0 +1,299 @@
+          }
+      }
+    }
+  }
+
+  def validateAndInsertDataset(dataset: Dataset): Int = {
+    var featuresWithoutIds = 0
+
+    var datasetStatus = 0
+    val assets = dataset.geoJson("features").asInstanceOf[List[Map[String, Any]]]
+    val roadlinks = dataset.roadlinks
+
+    if (assets.length != dataset.roadlinks.length) {
+      datasetStatus = 1
+    }
+
+    awsDao.insertDataset(dataset.datasetId, write(dataset.geoJson), write(dataset.roadlinks), datasetStatus)
+    if (datasetStatus != 1) {
+      var roadlinksCount = 0
+
+      assets.foreach(feature => {
+        val featureRoadlinks = roadlinks(roadlinksCount)
+        roadlinksCount += 1
+
+        val properties = feature.getOrElse("properties", "").asInstanceOf[Map[String, Any]]
+        properties.get("id") match {
+          case Some(featureId) =>
+            try {
+              awsDao.insertFeature(featureId.asInstanceOf[BigInt].longValue(), dataset.datasetId, 0)
+
+              linkIdValidation(featureRoadlinks.map(number => number.longValue()).toSet)
+
+              val assetTypeGeometry = feature("geometry").asInstanceOf[Map[String, Any]]("type")
+              assetTypeGeometry match {
+                case "LineString" =>
+                  properties("type").asInstanceOf[String] match {
+                    case "Roadlink" => validateRoadlink(dataset.datasetId, featureId.asInstanceOf[BigInt].longValue(), properties)
+                    case _ =>
+                      awsDao.updateFeatureStatus(featureId.asInstanceOf[BigInt].longValue(), "1")
+                      awsDao.updateDatasetStatus(dataset.datasetId, 2)
+                  }
+                case "Point" => validatePoint(dataset.datasetId, featureId.asInstanceOf[BigInt].longValue(), properties, properties("type").asInstanceOf[String])
+                case _ =>
+                  awsDao.updateFeatureStatus(featureId.asInstanceOf[BigInt].longValue(), "1")
+                  awsDao.updateDatasetStatus(dataset.datasetId, 2)
+              }
+            } catch {
+              case _: Throwable =>
+                awsDao.updateFeatureStatus(featureId.asInstanceOf[BigInt].longValue(), "1")
+                awsDao.updateDatasetStatus(dataset.datasetId, 2)
+            }
+          case None =>
+            featuresWithoutIds += 1
+            awsDao.updateDatasetStatus(dataset.datasetId, 2)
+        }
+      }
+      )
+    }
+    featuresWithoutIds
+  }
+
+  def updateDataset(dataset: Dataset) = {
+    if (awsDao.checkDatasetStatus(dataset.datasetId).toInt != 1) {
+      val assets = dataset.geoJson("features").asInstanceOf[List[Map[String, Any]]]
+      val roadlinks = dataset.roadlinks
+
+      var roadlinksCount = 0
+      assets.foreach(feature => {
+        val featureRoadlinks = roadlinks(roadlinksCount)
+        roadlinksCount += 1
+
+        val properties = feature("properties").asInstanceOf[Map[String, Any]]
+        val featureId = properties.getOrElse("id", 0)
+
+        if (featureId != 0) {
```

Comment:
update this featureId to be equal to the one used in validateAndInsertDataset method